File: java/src/main/java/org/nitk/hackathon/stock/StocksList.java
package org.nitk.hackathon.stock;

import org.nitk.hackathon.trader.Trader;

import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;

public class StocksList {

    private static StocksList instance;
    private List<Trader> traders = new ArrayList<>();
    private Map<String, Double> stocks = new HashMap<>();

    private StocksList() {
        // Initialize stocks with some initial values
        stocks.put("ABC", 100.0);
        stocks.put("XYZ", 50.0);
    }

    public static StocksList getObject() {    
        if (instance == null) {
            instance = new StocksList();
        }
        return instance;
    }

    public void registerTrader(Trader trader) {
        traders.add(trader);
    }

    public void updatePrice(String stock, double price) {
        stocks.put(stock, price);
        notifyTraders(stock, price);
    }

    private void notifyTraders(String stock, double price) {
        for (Trader trader : traders) {
            trader.update(stock, price);
        }
    }

}

File: python/stock/stocks_list.py
class StocksList:
    """
    A class representing a list of stocks and traders.
    """

    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(StocksList, cls).__new__(cls)
            cls._instance.__init__()
        return cls._instance

    def __init__(self):
        """
        Initialize the StocksList object.
        """
        self.traders = []
        self.stocks = {
            "ABC": 100.0,
            "XYZ": 50.0
        }

    def register_trader(self, trader):
        """
        Register a trader with the StocksList.
        """
        self.traders.append(trader)

    def update_price(self, stock, price):
        """
        Update the price of a stock and notify traders.
        """
        self.stocks[stock] = price
        self.notify_traders(stock, price)

    def notify_traders(self, stock, price):
        """
        Notify all registered traders about the stock price update.
        """
        for trader in self.traders:
            trader.update(stock, price)
